cdap-ranger/cdap-ranger-lookup/src/main/java/co/cask/cdap/security/authorization/ranger/lookup/client/CDAPClient.java

```diff
@@ -0,0 +1,226 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * CDAP Clients which to be used by ranger for listing cdap resources.
+ * Note: The log statements in this class are written in a formatted way to match the logging format
+ * used by other services in Apache Ranger since these log statements end up in ranger admin log file we
+ * want them to be in same format.
+ */
+public class CDAPClient {
+
+  private static final Log LOG = LogFactory.getLog(CDAPClient.class);
+  private static final long SERVICE_TIMEOUT_SECONDS = TimeUnit.MINUTES.toSeconds(30);
+
+  private AccessToken accessToken;
+  private final String serviceName;
+  private final String instanceURL;
+  private final String username;
+  private final String password;
+  private final NamespaceClient nsClient;
+  private static final String ERR_MSG = " You can still save the repository and start creating "
+    + "policies, but you would not be able to use autocomplete for "
+    + "resource names. Check xa_portal.log for more info.";
+
+  CDAPClient(String serviceName, String instanceURL, String username, String password) {
+    this.serviceName = serviceName;
+    this.instanceURL = instanceURL;
+    this.username = username;
+    this.password = password;
+    initConnection();
+    ClientConfig clientConfig = getClientConfig();
+    this.nsClient = new NamespaceClient(clientConfig);
+    //TODO create more cdap clients here
+  }
+
+  private void initConnection() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> CDAPClient initConnection()");
+    }
+    if (username != null && password != null) {
+      // security is enabled, we need to get access token before checking system services
+      try {
+        LOG.info(String.format("Fetching access token with username %s and password ****", username));
+        accessToken = fetchAccessToken(username, password);
+      } catch (Exception ex) {
+        throw Throwables.propagate(ex);
+      }
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("<== CDAPClient initConnection()");
+    }
+  }
+
+  /**
+   * Tests that a connection can be established to CDAP with the given config. We do this by listing namespaces.
+   * @return Map<String, Object> Connection test response
+   */
+  Map<String, Object> testConnection() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> CDAPClient testConnection()");
+    }
+    HashMap<String, Object> responseData = new HashMap<>();
+    List<String> testResult;
+    try {
+      testResult = getNamespaces(null);
+      if (testResult != null && testResult.size() >= 0) {
+        String successMsg = "Connection Test Successful";
+        BaseClient.generateResponseDataMap(true, successMsg, successMsg,
+                                           null, null, responseData);
+      } else {
```

Comment:
So the definition of testConnection() is the user can at least list one namespace. I am not sure if this is the correct way, since user can have privileges on just some entities other than namespace. Maybe we should consider the connection is successful if list() does not throw any exception?